install.sh

```diff
@@ -0,0 +1,176 @@
+#!/bin/sh
+
+# The original version of this script is licensed under the MIT license.
+# See https://github.com/Masterminds/glide/blob/master/LICENSE for more details
+# and copyright notice.
+
+PROJECT_NAME="arduino-cli"
+
+# BINDIR represents the local bin location, defaults to ./bin.
+LBINDIR=""
+DEFAULT_BINDIR="$PWD/bin"
+
+fail() {
+	echo "$1"
+	exit 1
+}
+
+initDestination() {
+	if [ -n "$BINDIR" ]; then
+		if [ ! -d "$BINDIR" ]; then
+			fail "$BINDIR "'($BINDIR)'" folder not found. Please create it before continue."
+		fi
+		LBINDIR="$BINDIR"
+	else
+		if [ ! -d "$DEFAULT_BINDIR" ]; then
+		mkdir "$DEFAULT_BINDIR"
+		fi
+		LBINDIR="$DEFAULT_BINDIR"
+	fi
+	echo "Installing in $LBINDIR"
+}
+
+initArch() {
+	ARCH=$(uname -m)
+	case $ARCH in
+		armv5*) ARCH="armv5";;
+		armv6*) ARCH="armv6";;
+		armv7*) ARCH="ARM";;
+		aarch64) ARCH="ARM64";;
+		x86) ARCH="32bit";;
+		x86_64) ARCH="64bit";;
+		i686) ARCH="32bit";;
+		i386) ARCH="32bit";;
+	esac
+	echo "ARCH=$ARCH"
+}
+
+initOS() {
+	OS=$(uname -s)
+	case "$OS" in
+		Linux*) OS='Linux' ;;
+		Darwin*) OS='macOS' ;;
+		MINGW*) OS='Windows';;
+		MSYS*) OS='Windows';;
+	esac
+	echo "OS=$OS"
+}
+
+initDownloadTool() {
+	if type "curl" > /dev/null; then
+		DOWNLOAD_TOOL="curl"
+	elif type "wget" > /dev/null; then
+		DOWNLOAD_TOOL="wget"
+	else
+		fail "You need curl or wget as download tool. Please install it first before continue"
+	fi
+	echo "Using $DOWNLOAD_TOOL as download tool"
+}
+
+get() {
+	local url="$2"
+	local body
+	local httpStatusCode
+	echo "Getting $url"
+	if [ "$DOWNLOAD_TOOL" = "curl" ]; then
+		httpResponse=$(curl -sL --write-out HTTPSTATUS:%{http_code} "$url")
+		httpStatusCode=$(echo $httpResponse | tr -d '\n' | sed -e 's/.*HTTPSTATUS://')
+		body=$(echo "$httpResponse" | sed -e 's/HTTPSTATUS\:.*//g')
+	elif [ "$DOWNLOAD_TOOL" = "wget" ]; then
+		tmpFile=$(mktemp)
+		body=$(wget --server-response --content-on-error -q -O - "$url" 2> $tmpFile || true)
+		httpStatusCode=$(cat $tmpFile | awk '/^  HTTP/{print $2}')
+	fi
+	if [ "$httpStatusCode" != 200 ]; then
+		echo "Request fail with http status code $httpStatusCode"
+		fail "Body: $body"
+	fi
+	eval "$1='$body'"
+}
+
+getFile() {
+	local url="$1"
+	local filePath="$2"
+	if [ "$DOWNLOAD_TOOL" = "curl" ]; then
+		httpStatusCode=$(curl -s -w '%{http_code}' -L "$url" -o "$filePath")
+	elif [ "$DOWNLOAD_TOOL" = "wget" ]; then
+		body=$(wget --server-response --content-on-error -q -O "$filePath" "$url")
+		httpStatusCode=$(cat $tmpFile | awk '/^  HTTP/{print $2}')
+	fi
+	echo "$httpStatusCode"
+}
+
+downloadFile() {
+	get TAG_JSON https://api.github.com/repos/arduino/arduino-cli/releases/latest
+	TAG=$(echo $TAG_JSON | python -c 'import json,sys;obj=json.load(sys.stdin);sys.stdout.write(obj["tag_name"])')
```

Comment:
We've noticed that the `eval "$1='$body'"` at line 88 turns the `\r\n` in the `body` property into real line feeds and carriage returns producing an error on this line 99:

```bash
Getting https://api.github.com/repos/arduino/arduino-cli/releases/latest
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Invalid control character at: line 1 column 13677 (char 13676)
```

As quick fix, since `\r` is not allowed in JSON, we've escaped this character in the string:

```suggestion
	TAG=$(echo $TAG_JSON | python -c 'import json,sys;data=sys.stdin.read();obj=json.loads(data.replace("\r", "\\r"));sys.stdout.write(obj["tag_name"])')
```

This works, but shouldn't we try a different solution changing the `get` function to not mess up with the JSON document?

Reply:
The `\r` seem to come from the `body` field - since we don't use it, can you try with `json.load(sys.stdin, strict=False)` without replacing the `\r` and see what happens?

Reply:
It works, it's one of the things we've tried before settle with the replace. So, let's go strictness!

Reply:
This is the fix using `strict=False`:

```suggestion
	TAG=$(echo $TAG_JSON | python -c 'import json,sys;obj=json.load(sys.stdin, strict=False);sys.stdout.write(obj["tag_name"])')
```